### DISCUSSION TITLE
v4 discussion

### DISCUSSION THREAD

**Discussion by nilsnolde** [valhalla/valhalla#3971] [Ideas]
We have a lot of new ideas how to burn bits, so I think this will become a dead-lock very soon, latest in summer we'll be out bits at least in directededge.

Discussing a bit the problem with @kevinkreiser , we feel it might be ok to get some bits back by "consolidating" logic so we'd be able to remove quite a few bits to open the project for more features. That's a lower-effort version of v4 without a bigger change as a whole.

I wasn't involved in the initial architecture process and wasn't there when all the big decisions were made, so I'm a bit reluctant to suggest anything without a deeper look into it. I (and my co-workers) can certainly contribute in getting there though. What do you think @dnesbitt61 @gknisely @dgearhart ? Would that sound like an acceptable way for you? If so, appreciated if you'd suggest stuff that could be removed off the top of your head, if any.

**Comment by dnesbitt61** [comment]
NodeInfo: 

- I've always thought that headings_ could be removed (outside of tile building I think it is only used in map-matching (could possibly be replaced with turn types / costs as in Costing methods?)

DirectedEdge:

- perhaps move specialized speeds (truck, free flow, constrained flow) into a parallel structure that is optional for those apps that don't require them.
- Density. I wonder if we can simplify this by assigning a density to each tile and using that density for any costing that needs it? Computing density per edge is costly (a high percentage of time spent in GraphEnhancer I think).
- Max upward and downward slope, I would remove these or refactor into a smaller number of bits. I think they are inaccurate anyway. Adding elevation to edges would allow this to be used (but not in routing computations).
- link (1 bit) - seems redundant (info contained in Use) and can likely be removed.

I am wondering if there is a way to allow building (and using) tile sets that support only a subset of access/attribution. For example, build a tile set that supports bicycle and pedestrian. There would then be many bits saved in the access (and restriction) fields which support 12 modes each.

**Reply by nilsnolde** [reply]
Thanks for your thoughts @dnesbitt61 !

> I am wondering if there is a way to allow building (and using) tile sets that support only a subset of access/attribution

Yeah, that'd be great, but sounds like it could be prohibiting building a graph for all modes in that case, IMO one of the (more or less) unique features of Valhalla. Also, that's likely more the kind of work which might be a better fit for a v5 (at least following the rationale of maximizing bit-to-effort relationship) :smile:

**Comment by nilsnolde** [comment]
Adding to the list:
- goodbye weird date format, hello ISO: the week starts on Monday! (really causing too many headaches and bugs in our code, as well as downstream, currently seeing it with GTFS/multimodal)
